### PR TITLE
r/aws_kinesis_firehose_delivery_stream add http_endpoint_configuration 

### LINKED ISSUE #14469: Add http_endpoint_configuration to r/aws_kinesis_firehose_delivery_stream
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Amazon Kinesis Data Firehose HTTP endpoint delivery was introduced to Firehose a week ago. This is a request to add `http_endpoint_configuration` to `aws_kinesis_firehose_delivery_stream`.

### New or Affected Resource(s)

r/aws_kinesis_firehose_delivery_stream

### Potential Terraform Configuration

```hcl
resource "aws_kinesis_firehose_delivery_stream" "test_stream" {
  name        = "terraform-kinesis-firehose-test-stream"
  destination = "http_endpoint"

  s3_configuration { // Similar to implementation in splunk_configuration, this maps to HttpEndpointDestinationConfiguration.S3Configuration
    role_arn           = aws_iam_role.firehose.arn
    bucket_arn         = aws_s3_bucket.bucket.arn
    buffer_size        = 10
    buffer_interval    = 400
    compression_format = "GZIP"
  }

  http_endpoint_configuration {
    url        = "https://api.example.com:443"
    name       = "MyHTTPEndpoint"
    access_key = "my-key"
    buffering_hints {
      size_in_mb = 5
      interval_in_seconds = 300
    }
   // cloudwatch_logging_options is CloudWatchLoggingOptions as used in s3_configuration
   // processing_configuration is ProcessingConfiguration as used in extended_s3_configuration
   role_arn       = "arn:my_role"
   s3_backup_mode = "FailedDataOnly"
  }
}
```

### References

* https://docs.aws.amazon.com/firehose/latest/APIReference/API_CreateDeliveryStream.html#Firehose-CreateDeliveryStream-request-HttpEndpointDestinationConfiguration
* https://aws.amazon.com/blogs/big-data/analyze-logs-with-datadog-using-amazon-kinesis-data-firehose-http-endpoint-delivery/

### PR DESCRIPTION
**Pull request by ZsoltPath**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #14469

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
resource/aws_kinesis_firehose_delivery_stream add http_endpoint_configuration 
```

### CODE REVIEW

**Comment by aldegoeij** [comment]
@bflad any luck reviewing this yet 😇

**Comment by saintberry** [comment]
Also super keen to see this get merged. It will be much easier to manage New Relic integration using Firehose vs Lambda.

**Review by bflad — CHANGES_REQUESTED** [review]
Hi @ZsoltPath 👋  Thank you for submitting this! At quick glance, this looks like its on the right track here for the code and documentation, however we will need to also ensure there is testing added to verify the functionality. Is this something you are able or interested to work on?

If so, for more information about how the testing works in this project, please see the [Running and Writing Acceptance Testing section of the Contributing Guide](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/contributing/running-and-writing-acceptance-tests.md). Copy-paste-modify of existing testing is fine and encouraged. The maintainers can try to answer any questions you might have.

If not, no big deal, either someone from the community can pick it up or eventually a maintainer.

Please let us know, thanks!

**Comment by ZsoltPath** [comment]
> 
> 
> Hi @ZsoltPath 👋 Thank you for submitting this! At quick glance, this looks like its on the right track here for the code and documentation, however we will need to also ensure there is testing added to verify the functionality. Is this something you are able or interested to work on?
> 
> If so, for more information about how the testing works in this project, please see the [Running and Writing Acceptance Testing section of the Contributing Guide](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/contributing/running-and-writing-acceptance-tests.md). Copy-paste-modify of existing testing is fine and encouraged. The maintainers can try to answer any questions you might have.
> 
> If not, no big deal, either someone from the community can pick it up or eventually a maintainer.
> 
> Please let us know, thanks!

I'll give it a go.
I wasn't super keen on it because as I see very few of the resource's functions have tests already.

**Comment by ZsoltPath** [comment]
@bflad I'm stuck here with HCL formatting.
When I run terrafmt locally it passes without an issue:
```
terraform-provider-aws$ terrafmt diff --check --fmtcompat aws/resource_aws_kinesis_firehose_delivery_stream_test.go
terraform-provider-aws$ echo $?
0
```

Could you give me a hint what should I do? I just don't want to play around and push commits and commits to accidentally find the solution.

File: aws/resource_aws_kinesis_firehose_delivery_stream_test.go
Comment (bflad) [inline aws/resource_aws_kinesis_firehose_delivery_stream_test.go:2736]:
```suggestion
}
```

File: aws/resource_aws_kinesis_firehose_delivery_stream_test.go
Comment (bflad) [inline aws/resource_aws_kinesis_firehose_delivery_stream_test.go:2768]:
```suggestion
          parameter_name  = "LambdaArn"
          parameter_value = "${aws_lambda_function.lambda_function_test.arn}:$LATEST"
```

File: aws/resource_aws_kinesis_firehose_delivery_stream_test.go
Comment (bflad) [inline aws/resource_aws_kinesis_firehose_delivery_stream_test.go:2773]:
```suggestion
          parameter_name  = "RoleArn"
          parameter_value = aws_iam_role.firehose.arn
```

File: aws/resource_aws_kinesis_firehose_delivery_stream_test.go
Comment (bflad) [inline aws/resource_aws_kinesis_firehose_delivery_stream_test.go:2778]:
```suggestion
          parameter_name  = "BufferSizeInMBs"
          parameter_value = 1
```

File: aws/resource_aws_kinesis_firehose_delivery_stream_test.go
Comment (bflad) [inline aws/resource_aws_kinesis_firehose_delivery_stream_test.go:2788]:
```suggestion
}
```

File: aws/resource_aws_kinesis_firehose_delivery_stream_test.go
Comment (bflad) [inline aws/resource_aws_kinesis_firehose_delivery_stream_test.go:2829]:
```suggestion
    access_key     = "test_key"
    role_arn       = aws_iam_role.firehose.arn
    s3_backup_mode = "FailedEventsOnly"
```

File: aws/resource_aws_kinesis_firehose_delivery_stream_test.go
Comment (bflad) [inline aws/resource_aws_kinesis_firehose_delivery_stream_test.go:2855]:
```suggestion
        parameters {
          parameter_name  = "LambdaArn"
          parameter_value = "${aws_lambda_function.lambda_function_test.arn}:$LATEST"
        }

        parameters {
          parameter_name  = "RoleArn"
          parameter_value = aws_iam_role.firehose.arn
        }

        parameters {
          parameter_name  = "BufferSizeInMBs"
          parameter_value = 1
        }

        parameters {
          parameter_name  = "BufferIntervalInSeconds"
          parameter_value = 120
        }
```

File: aws/resource_aws_kinesis_firehose_delivery_stream_test.go
Comment (bflad) [inline aws/resource_aws_kinesis_firehose_delivery_stream_test.go:2859]:
```suggestion
}
```

**Review by bflad — COMMENTED** [review]
I'll need to ask around why `terrafmt` isn't working locally as I'm not able to reproduce the difference output myself, even though there definitely was whitespace issues. I tried to catch the ones I could see though and maybe this will square it away.

**Comment by ZsoltPath** [comment]
> 
> 
> I'll need to ask around why `terrafmt` isn't working locally as I'm not able to reproduce the difference output myself, even though there definitely was whitespace issues. I tried to catch the ones I could see though and maybe this will square it away.

Yay! Thanks for finding these! I hate whitespace issues.

File: aws/resource_aws_kinesis_firehose_delivery_stream_test.go
Comment (bflad) [inline aws/resource_aws_kinesis_firehose_delivery_stream_test.go:2813]: The acceptance test is currently failing name length validation:

```
=== CONT  TestAccAWSKinesisFirehoseDeliveryStream_HTTPEndpointConfigUpdates
    resource_aws_kinesis_firehose_delivery_stream_test.go:1174: Step 1/3 error: Error running pre-apply refresh: 2020/11/05 20:03:53 [DEBUG] Using modified User-Agent: Terraform/0.12.29 HashiCorp-terraform-exec/0.10.0

        Error: expected length of name to be in the range (1 - 64), got terraform-kinesis-firehose-basichttpendpointtest-5506598618522017619

          on ../../../../../../../../var/folders/w8/05f3x02n27x72g0mc2jy6_180000gp/T/plugintest966219592/work600651783/config498403883/terraform_plugin_test.tf line 77, in resource "aws_kinesis_firehose_delivery_stream" "test":
          77:   name        = "terraform-kinesis-firehose-basichttpendpointtest-5506598618522017619"


--- FAIL: TestAccAWSKinesisFirehoseDeliveryStream_HTTPEndpointConfigUpdates (1.83s)
```

```suggestion
  name        = "terraform-kinesis-firehose-httpendpoint-%d"
  destination = "http_endpoint"

  s3_configuration {
    role_arn   = aws_iam_role.firehose.arn
    bucket_arn = aws_s3_bucket.bucket.arn
  }

  http_endpoint_configuration {
    url      = "https://input-test.com:443"
    name     = "HTTP_test"
    role_arn = aws_iam_role.firehose.arn
  }
}
`

var testAccKinesisFirehoseDeliveryStreamConfig_HTTPEndpointUpdates = testAccKinesisFirehoseDeliveryStreamBaseConfig + `
resource "aws_kinesis_firehose_delivery_stream" "test" {
  depends_on  = [aws_iam_role_policy.firehose]
  name        = "terraform-kinesis-firehose-httpendpoint-%d"
```

File: aws/resource_aws_kinesis_firehose_delivery_stream.go
Comment (bflad) [inline aws/resource_aws_kinesis_firehose_delivery_stream.go:1497]: This attribute does not appear to be handled in the expand functions or documented. When running the current acceptance test with fixing the name length, the API seems to be setting this to 300 by default with it unconfigured:

```
=== CONT  TestAccAWSKinesisFirehoseDeliveryStream_HTTPEndpointConfigUpdates
    resource_aws_kinesis_firehose_delivery_stream_test.go:1174: Step 1/3 error: After applying this test step, the plan was not empty.
        stdout:


        An execution plan has been generated and is shown below.
        Resource actions are indicated with the following symbols:
          ~ update in-place

        Terraform will perform the following actions:

          # aws_kinesis_firehose_delivery_stream.test will be updated in-place
          ~ resource "aws_kinesis_firehose_delivery_stream" "test" {
                arn            = "arn:aws:firehose:us-west-2:--OMITTED--:deliverystream/terraform-kinesis-firehose-httpendpoint-1205375700746293683"
                destination    = "http_endpoint"
                destination_id = "destinationId-000000000001"
                id             = "arn:aws:firehose:us-west-2:--OMITTED--:deliverystream/terraform-kinesis-firehose-httpendpoint-1205375700746293683"
                name           = "terraform-kinesis-firehose-httpendpoint-1205375700746293683"
                version_id     = "1"

              ~ http_endpoint_configuration {
                    buffering_interval = 300
                    buffering_size     = 5
                    name               = "HTTP_test"
                  ~ retry_duration     = 300 -> 3600
                    role_arn           = "arn:aws:iam::--OMITTED--:role/tf_acctest_firehose_delivery_role_1205375700746293683"
                    s3_backup_mode     = "FailedDataOnly"
                    url                = "https://input-test.com:443"

                    cloudwatch_logging_options {
                        enabled = false
                    }

                    processing_configuration {
                        enabled = false
                    }

                    request_configuration {
                        content_encoding = "NONE"
                    }
                }

                s3_configuration {
                    bucket_arn         = "arn:aws:s3:::tf-test-bucket-1205375700746293683"
                    buffer_interval    = 300
                    buffer_size        = 5
                    compression_format = "UNCOMPRESSED"
                    role_arn           = "arn:aws:iam::--OMITTED--:role/tf_acctest_firehose_delivery_role_1205375700746293683"

                    cloudwatch_logging_options {
                        enabled = false
                    }
                }

                server_side_encryption {
                    enabled  = false
                    key_type = "AWS_OWNED_CMK"
                }
            }

        Plan: 0 to add, 1 to change, 0 to destroy.
--- FAIL: TestAccAWSKinesisFirehoseDeliveryStream_HTTPEndpointConfigUpdates (94.61s)
```

```suggestion
						"retry_duration": {
							Type:         schema.TypeInt,
							Optional:     true,
							Default:      300,
							ValidateFunc: validation.IntBetween(0, 7200),
						},
```

**Review by bflad — CHANGES_REQUESTED** [review]
Hi @ZsoltPath 👋  Getting close. Are you able to run the acceptance testing? Please see the [Running and Writing Acceptance Testing section of the Contributing Guide](https://github.com/hashicorp/terraform-provider-aws/blob/master/docs/contributing/running-and-writing-acceptance-tests.md). The other resource tests look fine, just need to fix up this one. 👍

**Comment by ZsoltPath** [comment]
How the heck this new semgrep issue come up for `aws/resource_aws_s3_bucket_public_access_block.go`?
I haven't even opened that file :(

**Comment by bflad** [comment]
@ZsoltPath it is unrelated to your change and is due to merging https://github.com/hashicorp/terraform-provider-aws/pull/15925. It was fixed earlier today with https://github.com/hashicorp/terraform-provider-aws/pull/16068. `semgrep` is supposed to be better about ignoring unrelated changes in pull requests by comparing previous Git history, but seems like that isn't working correctly. Regardless, that issue will not affect me reviewing and potentially merging this if the acceptance testing is now working.

File: aws/resource_aws_kinesis_firehose_delivery_stream.go
Comment (bflad) [reply aws/resource_aws_kinesis_firehose_delivery_stream.go]: This attribute is still not handled in the expand functions or documented. Will handle on merge to save the back and forth.

**Review by bflad — COMMENTED** [review]
Thank you for your work on this, @ZsoltPath. 👍  It is close enough that I'm going to squash and merge this to fix up the pending test failure and ensure that `retry_duration` is properly configurable.

**Pull request merged**
